### PR TITLE
Port to `System.Threading.Channels`

### PR DESCRIPTION
**Pull request by nblumhardt**
This is effectively a complete rewrite; it's intended to be more efficient, robust, and modern, but it's going to need some bake time.

The sink now uses _System.Threading.Channels_ to implement an async queue, so there's no longer a blocked worker thread waiting to process batches.

The sink also now begins processing batches as soon as they're full, so if the batch size is 100, and 100 events are written, they'll be processed regardless of whether the buffering delay (`PeriodicBatchingSinkOptions.Period`) has elapsed.

The sink now properly avoids overlapping batches (the old `Monitor`-based logic was/is broken, as far as I can tell). Shut-down/flush logic is also cleaner: the queue will always be fully-flushed on shutdown, except when an error occurs.

Finally:

 * The internals of the sink are now probably close enough to what would be required by _Serilog.Sinks.Async_ that the two could look at converging in the future (`WriteTo.Batched(wt => ...)` is also a reasonable thing to consider doing with the new API)
 * The internals could now support a concurrency level `>  1` by kicking off multiple instances of the run loop, if throughput requires it

I've dropped out the old performance tests, which only effectively covered the custom blocking collection type that the sink used. I'll explore performance using downstream sinks and try to formulate a better perf plan.

**Breaking changes:**

 * TFM support has been narrowed to effectively only targets supported by `netstandard2.0`; version 3.x of this project can continue supporting older/obsolete TFMs
 * The inheritance-based mode is gone; consumers always implement `IBatchedLogEventSink` and instantiate `PeriodicBatchingSink` as a wrapper
 * Queue size matters more, as it affects shut-down flush time
 * Batching logic has changed, so performance may vary in unexpected ways from the older version

### CODE REVIEW

File: .DS_Store
Comment (bartelink) [inline .DS_Store:1]: remove + .gitignore

File: src/Serilog.Sinks.PeriodicBatching/Sinks/PeriodicBatching/PeriodicBatchingSink.cs [hunk 1]
```diff
@@ -23,31 +24,26 @@ namespace Serilog.Sinks.PeriodicBatching;
...
 {
-    /// <summary>
-    /// Constant used with legacy constructor to indicate that the internal queue shouldn't be limited.
-    /// </summary>
-    [Obsolete("Implement `IBatchedLogEventSink` and use the `PeriodicBatchingSinkOptions` constructor.")]
-    public const int NoQueueLimit = -1;
+    // Buffers events from the write- to the read side.
+    readonly Channel<LogEvent> _queue;
 
+    // Used by the write side to signal shutdown.
```
Comment (bartelink) [inline src/Serilog.Sinks.PeriodicBatching/Sinks/PeriodicBatching/PeriodicBatchingSink.cs]: maybe a note as to why marking the channel closed can't be used to convey that?
(Only scanning, could be way off)

File: src/Serilog.Sinks.PeriodicBatching/Serilog.Sinks.PeriodicBatching.csproj [hunk 2]
```diff
@@ -2,9 +2,10 @@
 
   <PropertyGroup>
     <Description>Buffer batches of log events to be flushed asynchronously.</Description>
-    <VersionPrefix>3.1.1</VersionPrefix>
+    <VersionPrefix>4.0.0</VersionPrefix>
     <Authors>Serilog Contributors</Authors>
-    <TargetFrameworks>net45;netstandard1.1;netstandard1.2;netstandard2.0;netstandard2.1</TargetFrameworks>
+    <TargetFrameworks Condition=" '$(OS)' == 'Windows_NT'">net462</TargetFrameworks>
+    <TargetFrameworks>$(TargetFrameworks);netstandard2.0;netstandard2.1;net6.0</TargetFrameworks>
```
Comment (bartelink) [inline src/Serilog.Sinks.PeriodicBatching/Serilog.Sinks.PeriodicBatching.csproj]: why retaining NS2.1? (Struggling to recall a good reason? https://github.com/serilog/serilog/issues/1970#issuecomment-1814286393)

**Comment by nblumhardt** [comment]
Thanks @bartelink! Currently just roughing it out - will loop back and make sure I've sorted those things out once I've decided on how shutdown should work 👍

**Comment by nblumhardt** [comment]
Tests are failing on .NET Framework (yay 🙄), will sort that out shortly, and then with some luck this will be ready for a look! :-)

File: test/Serilog.Sinks.PeriodicBatching.Tests/Serilog.Sinks.PeriodicBatching.Tests.csproj [hunk 3]
```diff
@@ -1,14 +1,14 @@
...
+    <TargetFrameworks Condition=" '$(OS)' == 'Windows_NT'">net472</TargetFrameworks>
+    <TargetFrameworks>$(TargetFrameworks);net8.0</TargetFrameworks>
     <AssemblyOriginatorKeyFile>../../assets/Serilog.snk</AssemblyOriginatorKeyFile>
     <SignAssembly>true</SignAssembly>
     <PublicSign Condition=" '$(OS)' != 'Windows_NT' ">true</PublicSign>
-    <CheckEolTargetFramework>false</CheckEolTargetFramework>
   </PropertyGroup>
 
-  <PropertyGroup Condition=" '$(TargetFramework)' == 'net6.0' ">
+  <PropertyGroup Condition=" '$(TargetFramework)' == 'net8.0' ">
```
Comment (bartelink) [inline test/Serilog.Sinks.PeriodicBatching.Tests/Serilog.Sinks.PeriodicBatching.Tests.csproj:11]: this seems redundant now unless you plan to do some testing via it?

File: src/Serilog.Sinks.PeriodicBatching/Sinks/PeriodicBatching/PeriodicBatchingSink.cs [hunk 4]
```diff
@@ -56,105 +58,72 @@ public class PeriodicBatchingSink : ILogEventSink, IDisposable, IBatchedLogEvent
...
+        try
         {
-            if (_unloading)
-                return;
+            lock (_stateLock)
+            {
+                if (!_unloading.IsCancellationRequested)
+                {
+                    _queue.Writer.Complete();
+                    _unloading.Cancel();
```
Comment (bartelink) [inline src/Serilog.Sinks.PeriodicBatching/Sinks/PeriodicBatching/PeriodicBatchingSink.cs]: Perhaps shutdownCts or something might be a better name? It took me till this far down the code despite a decent comment over it above

Comment (nblumhardt) [reply src/Serilog.Sinks.PeriodicBatching/Sinks/PeriodicBatching/PeriodicBatchingSink.cs]: Had a shot at some renames across the board, thanks for the feedback 👍

**Comment by nblumhardt** [comment]
I think this is ready for a closer look now :-)

**Comment by nblumhardt** [comment]
@skomis-mm hope all is well with you! Just flagging this as one that would benefit from your eyes, in case you're able to take a look at any point.

Planning to merge fairly soon, to keep pushing ahead, but expecting to run through a few iterations before calling it "stable".

File: src/Serilog.Sinks.PeriodicBatching/Sinks/PeriodicBatching/PeriodicBatchingSink.cs [hunk 5]
```diff
@@ -56,305 +58,209 @@ public class PeriodicBatchingSink : ILogEventSink, IDisposable, IBatchedLogEvent
...
+    }
+    
+    /// <inheritdoc/>
+    public void Dispose()
+    {
+        try
         {
-            if (_unloading)
-                return;
+            lock (_stateLock)
```
Comment (bartelink) [inline src/Serilog.Sinks.PeriodicBatching/Sinks/PeriodicBatching/PeriodicBatchingSink.cs]: Extract L107-114  to common helper (to make the Async cs Sync diferences clearer)

File: src/Serilog.Sinks.PeriodicBatching/Sinks/PeriodicBatching/PeriodicBatchingSink.cs [hunk 6]
```diff
@@ -56,305 +58,209 @@ public class PeriodicBatchingSink : ILogEventSink, IDisposable, IBatchedLogEvent
...
-        if (!disposing) return;
+        if (_shutdownSignal.IsCancellationRequested)
+            return;
 
-        lock (_stateLock)
+        _queue.Writer.TryWrite(logEvent);
+    }
+    
+    /// <inheritdoc/>
+    public void Dispose()
```
Comment (bartelink) [inline src/Serilog.Sinks.PeriodicBatching/Sinks/PeriodicBatching/PeriodicBatchingSink.cs]: I know the diffs will look bad snd it prob does not match the style of other sinks, but the Dispose stuff needs to move out of here to the end!
aka have the core impl near all the supporting state vars

File: src/Serilog.Sinks.PeriodicBatching/Sinks/PeriodicBatching/PeriodicBatchingSink.cs [hunk 7]
```diff
@@ -56,305 +58,209 @@ public class PeriodicBatchingSink : ILogEventSink, IDisposable, IBatchedLogEvent
...
+                    isEagerBatch = false;
+                    await _targetSink.EmitBatchAsync(_currentBatch).ConfigureAwait(false);
                 }
 
-                await _batchedLogEventSink.EmitBatchAsync(_waitingBatch).ConfigureAwait(false);
-
-                batchWasFull = _waitingBatch.Count >= _batchSizeLimit;
-                _waitingBatch.Clear();
-                _status.MarkSuccess();
+                _currentBatch.Clear();
```
Comment (bartelink) [inline src/Serilog.Sinks.PeriodicBatching/Sinks/PeriodicBatching/PeriodicBatchingSink.cs]: up into the else

File: src/Serilog.Sinks.PeriodicBatching/Sinks/PeriodicBatching/PeriodicBatchingSink.cs [hunk 8]
```diff
@@ -56,305 +58,209 @@ public class PeriodicBatchingSink : ILogEventSink, IDisposable, IBatchedLogEvent
...
+                if (_batchScheduler.ShouldDropBatch)
+                    _currentBatch.Clear();
 
-            lock (_stateLock)
-            {
-                if (!_unloading)
-                    SetTimer(_status.NextInterval);
+                if (_batchScheduler.ShouldDropQueue)
+                {
+                    // This is not ideal; the goal is to reduce memory pressure on the client if the server is offline
```
Comment (bartelink) [inline src/Serilog.Sinks.PeriodicBatching/Sinks/PeriodicBatching/PeriodicBatchingSink.cs]: this should at least selflog (maybe also when buffer is full and TryWrite is refused?)

File: test/Serilog.Sinks.PeriodicBatching.Tests/FailureAwareBatchSchedulerTests.cs [hunk 9]
```diff
@@ -3,29 +3,29 @@
...
-public class BatchedConnectionStatusTests
+public class FailureAwareBatchSchedulerTests
 {
     readonly TimeSpan _defaultPeriod = TimeSpan.FromSeconds(2);
 
     [Fact]
     public void WhenNoFailuresHaveOccurredTheRegularIntervalIsUsed()
     {
-        var bcs = new BatchedConnectionStatus(_defaultPeriod);
+        var bcs = new FailureAwareBatchScheduler(_defaultPeriod);
```
Comment (bartelink) [inline test/Serilog.Sinks.PeriodicBatching.Tests/FailureAwareBatchSchedulerTests.cs]: move this to L9 and remove from all other bodies

**Comment by nblumhardt** [comment]
Thanks for the awesome review @bartelink 👍 should all be done, now.

File: src/Serilog.Sinks.PeriodicBatching/Sinks/PeriodicBatching/PeriodicBatchingSink.cs [hunk 10]
```diff
@@ -187,24 +130,31 @@ async Task LoopAsync()
...
 
-                _currentBatch.Clear();
-                _batchScheduler.MarkSuccess();
+                    _currentBatch.Clear();
+                    _batchScheduler.MarkSuccess();
+                }
             }
             catch (Exception ex)
             {
                 SelfLog.WriteLine($"PeriodicBatchingSink ({_targetSink}) failed emitting a batch: {ex}");
```
Comment (bartelink) [inline src/Serilog.Sinks.PeriodicBatching/Sinks/PeriodicBatching/PeriodicBatchingSink.cs]: It's definitely looking nice and legible to my eyes now

Perhaps all the selflog emissions could be managed as a block? Maybe as virtuals in a base class? if there was one per case, maybe they could have header comments with explanations as to why seflogging without any way to hook it is the chosen mechanism?

Comment (nblumhardt) [reply src/Serilog.Sinks.PeriodicBatching/Sinks/PeriodicBatching/PeriodicBatchingSink.cs]: Factored out but kept it pretty simple, I think eventually we'll be able to deal with more interesting use cases via the queue monitoring and (hypothetical) fallback mechanisms.

Will hit the button, CI is green :-) 🎉

**Comment by nblumhardt** [comment]
Published to NuGet now in `4.0.0-dev-*`

**Pull request merged**